pkg/controller/webhooks/keycloak.go

```diff
@@ -0,0 +1,78 @@
+package webhooks
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	vshnv1 "github.com/vshn/appcat/v4/apis/vshn/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+//+kubebuilder:webhook:verbs=delete,path=/validate-vshn-appcat-vshn-io-v1-vshnkeycloak,mutating=false,failurePolicy=fail,groups=vshn.appcat.vshn.io,resources=vshnkeycloaks,versions=v1,name=vshnkeycloak.vshn.appcat.vshn.io,sideEffects=None,admissionReviewVersions=v1
+
+//+kubebuilder:rbac:groups=vshn.appcat.vshn.io,resources=xvshnkeycloaks,verbs=get;list;watch;patch;update
+//+kubebuilder:rbac:groups=vshn.appcat.vshn.io,resources=xvshnkeycloaks/status,verbs=get;list;watch;patch;update
+
+var (
+	keycloakGK = schema.GroupKind{Group: "vshn.appcat.vshn.io", Kind: "VSHNKeycloak"}
+)
+
+var _ webhook.CustomValidator = &KeycloakWebhookHandler{}
+
+type KeycloakWebhookHandler struct {
+	client    client.Client
+	log       logr.Logger
+	withQuota bool
+}
+
+// SetupKeycloakWebhookHandlerWithManager registers the validation webhook with the manager.
+func SetupKeycloakWebhookHandlerWithManager(mgr ctrl.Manager, withQuota bool) error {
+
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(&vshnv1.VSHNKeycloak{}).
+		WithValidator(&KeycloakWebhookHandler{
+			client:    mgr.GetClient(),
+			log:       mgr.GetLogger().WithName("webhook").WithName("keycloak"),
+			withQuota: withQuota,
+		}).
+		Complete()
+}
+
+// ValidateCreate implements webhook.CustomValidator so a webhook will be registered for the type
+func (r *KeycloakWebhookHandler) ValidateCreate(ctx context.Context, obj runtime.Object) (admission.Warnings, error) {
+	return nil, nil
+}
+
+// ValidateUpdate implements webhook.CustomValidator so a webhook will be registered for the type
+func (r *KeycloakWebhookHandler) ValidateUpdate(ctx context.Context, oldObj, newObj runtime.Object) (admission.Warnings, error) {
+	return nil, nil
+}
+
+// ValidateDelete implements webhook.CustomValidator so a webhook will be registered for the type
+func (r *KeycloakWebhookHandler) ValidateDelete(ctx context.Context, obj runtime.Object) (admission.Warnings, error) {
```

Comment:
I just realized that this function takes a `runtime.Object`. There's potential to generalize this away as well.

We could have a default webhook handler that implements this validation function. But instead of casting to concrete a concrete type, we create a new interface that returns the value of the boolean (`ProtectionChecker` or something), that we can use to cast to.

Then each of the concrete handlers we have can "inherit" this default handler and override the functions if necessary.  A similar approach has been used with the API Server: https://github.com/vshn/appcat/blob/master/pkg/apiserver/vshn/mariadb/backup.go#L25. Here you can see that the `noop.Noop` handler is an anonymous field of the `vshnMariaDBBackupStorage` handler. Which acts somewhat like inheritance in other languages.

There are various pros to this:
* Get rid of the same code over and over
* The webhook template for new services can be much smaller
* We have the code once and if it needs changes we need to change it once for all services
* Adding a new service to the webhooks is easy as "inheriting" the default handler already does 99% of the work for you

Reply:
I've refactored the code as discussed